### ISSUE TITLE
File Store needed

### ISSUE THREAD

**Issue by nicholascar** [RDFLib/rdflib#1207]
Currently RDFlib has a `Store` for in-memory data, SPARQL endpoints (read-only & read/write) and other specialised DBs but it doesn't have a general-purpose _file_ `Store`. 

Such a `Store` would probably:

* parse RDF files in a directory into an in-memory store and keep the files/in-memory object in sync

To do this it would likely:

* watch a directory
* on file change
    * either reload the entire in-memory graph or
    * reload one graph in an in-memory `Dataset` for just the file that changed

It would be hard for it to handle graph-aware files, so these might be excluded in v1 of such a store.

**Comment by white-gecko** [comment]
There is https://github.com/nareike/adhs by @nareike and a fork by @splattater who extended it to also support quads/a graph-aware store: https://github.com/splattater/adhs/tree/feature/WSGI+QUADS

@splattater and have then built the QuitStore (https://github.com/AKSW/QuitStore) which is a read/write graph-aware store that also versions all of the data in Git.

For sure I would like to see improvements to the file system side and we could modularize it to make it a proper RDFlib store.

**Comment by tgbugs** [comment]
I don't think this is how stores are intended to work. What is described in the requirements would be some other process that interacts with a memory store. If we were to follow the usual meaning of *Store in rdflib then a FileStore would be a store that
was capable of performing the usual rdflib operations without loading the whole graph into memory, basically it would be a DoesNotExistStore since iterating over triples would go straight from the parser out the generator without stopping anywhere inbetween.

Maybe what is described in the OP could be called a SyncMemoryStore or something like that?

**Comment by nicholascar** [comment]
> What is described in the requirements would be some other process that interacts with a memory store

Agree, this speaks more to the requirements.

> could be called a SyncMemoryStore or something like that

Perhaps you're right here @tgbugs. Perhaps even a FileSyncMemoryStore or a FilePickleStore

Clearly I've not worked out all the modes of operation here and the main conceptual issue (for me) still is whether or not it's still a "Store" if it persists between Python (RDFlib) sessions. This "Store" may:

- cache RDF files as Pickles and sync new/changed files to Pickles
- Allow an active (in session) Memory Store to be updated if any changes to the files (and thus the Pickles) takes place

Perhaps we (@jamiefeiss & I) need to test out a few usage scenarios...
